ENIGMAsystem/SHELL/Graphics_Systems/General/GSmodel_impl.h

```diff
@@ -0,0 +1,56 @@
+/** Copyright (C) 2013 Robert Colton, Adriano Tumminelli
+*** Copyright (C) 2014 Seth N. Hetu, Josh Ventura, Harijs Grinbergs
+*** Copyright (C) 2015 Harijs Grinbergs
+*** Copyright (C) 2018 Robert Colton
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#ifdef INCLUDED_FROM_SHELLMAIN
+#  error This file includes non-ENIGMA STL headers and should not be included from SHELLmain.
+#endif
+
+#ifndef ENIGMA_GSMODEL_IMPL_H
+#define ENIGMA_GSMODEL_IMPL_H
+
+#include <vector>
+using std::vector;
+
+namespace enigma {
+
+struct Primitive {
+  int type, format;
+  bool format_defined;
+  unsigned vertex_start, vertex_count;
+
+  Primitive(int type, int format, bool format_defined, unsigned start):
+    type(type), format(format), format_defined(format_defined), vertex_start(start), vertex_count(0) {}
+};
+
+struct Model {
+  int type, vertex_buffer;

```

Comment:
Sure: af21dc3e89176d30a18d93aceb9f0942e3ed7165

And no, it's not going to leak, I was very careful to ensure it doesn't and the benchmark shows that. I specified in the contract that the life of the vertex buffer should be the same as the lifetime of the model.

However, I should mention that GMS does not use regular ids for `vertex_*` functions. They use a pointer just as you are suggesting, which we can do now since you helped me overload var with `is_ptr()` and all that crap. Regardless, I chose to use normal integer-based ids for `vertex_create_buffer` because it's more consistent with GM's existing API. I see no reason to introduce a consistency here and I don't think our users care because they won't notice.
https://docs.yoyogames.com/source/dadiospice/002_reference/shaders/primitive%20building/vertex_create_buffer.html

![Vertex Buffers in GMS are Pointers](https://user-images.githubusercontent.com/3212801/44075965-8706e258-9f6c-11e8-9072-eab97b8c8637.png)

What do you think?